src/client/datascience/interactive-window/nativeInteractiveWindowProvider.ts

```diff
@@ -209,6 +243,27 @@ export class NativeInteractiveWindowProvider implements IInteractiveWindowProvid
         });
     }
 
+    private update() {
+        const windows: NativeInteractiveWindow[] = [];
+
+        this._windows.forEach((win) => {
+            const notebookDocument = workspace.notebookDocuments.find(
```

Comment:
Not sure this is possible.
If we're holding onto a reference of a Notebook, how is it that it that we're holding onto a reference of that object when this notebook no longer exists.

I understand its a simple block of code, but I'd prefer to avoid such code that account for scenarios that should tehcnically never occur.
I.e. if we handle the onDidCloseNotebook document, then the interactive window would get diposed.

Reply:
Also I think it might be better to use expose the `NotebookDocument isntead of `NotebookUri` as you're comparing Uris, when its more efficient to compare the object reference.

Reply:
> if we handle the onDidCloseNotebook document, then the interactive window would get diposed.

That's exactly what this code is doing. This is the function that gets called when `workspace.onDidCloseNotebookDocument` fires, and it ensures that we dispose our NativeInteractiveWindow wrapper objects if the underlying NotebookDocument has been disposed (which can absolutely happen).

Reply:
Maybe you're suggesting that inside the NativeInteractiveWindow, we should listen for onDidCloseNotebookDocument, and call dispose on ourselves if the closed NotebookDocument is the one we hold a reference to? That's how I originally implemented it, but Peng and I discussed and agreed it seemed weird for an object to be disposing itself. If anything an object's manager/provider should be handling disposal. That's why we landed on this pattern.

Reply:
> listen for onDidCloseNotebookDocument, and call dispose on ourselves if the closed NotebookDocument is the on

Yes, thats right.

> to be disposing itself. If anything an object's manager/provider should be handling disposal. That's why we landed on this pattern.

But then why can't we have a manager listen to the dispose event and then dispose the interactive windows. 
Anyways, looks like this has been discussed, lets leave this as is then.